packages/server/src/fhir/operations/resourcegraph.ts

```diff
@@ -0,0 +1,333 @@
+async function followLinks(
+  resource: Resource,
+  links: GraphDefinitionLink[] | undefined,
+  results: Resource[],
+  resourceCache: Record<string, Resource>,
+  repo: Repository,
+  depth: number = 0
+) {
+  // Circuit Breaker
+  if (results.length > 1000 || depth >= 5) {
+    return;
+  }
+
+  // Recursive Base Case
+  if (!links || links.length == 0) {
+    return;
+  }
+  for (const link of links) {
+    if (!link.target) {
+      continue;
+    }
+    for (const target of link.target) {
+      let linkedResources = [] as Resource[];
+
+      if (isFhirPathLink(link)) {
+        linkedResources = await followFhirPathLink(link, target, resource, resourceCache, repo);
+      } else if (isSearchLink(link)) {
+        linkedResources = await followSearchLink(resource, link, resourceCache, repo);
+      } else {
+        throw new OperationOutcomeError(badRequest(`Invalid link: ${JSON.stringify(link)}`));
+      }
+      linkedResources = deduplicateResources(linkedResources);
+      results.push(...linkedResources);
+      for (const linkedResource of linkedResources) {
+        await followLinks(linkedResource, target.link, results, resourceCache, repo, depth + 1);
+      }
+    }
+  }
+}
+
+/**
+ *
+ * If link.path:
+ *  Get elements
+ *  For target in Targets
+ *    For elem in elements:
+ *      If element is reference: follow reference
+ *      If element is canonical: search for resources with url===canonical
+ * @param link
+ * @param resource
+ * @param repo
+ * @param results
+ */
+async function followFhirPathLink(
+  link: FhirPathLink,
+  target: GraphDefinitionLinkTarget,
+  resource: Resource,
+  resourceCache: Record<string, Resource>,
+  repo: Repository
+): Promise<Resource[]> {
+  const results = [] as Resource[];
+
+  const elements = evalFhirPathTyped(link.path, [toTypedValue(resource)]);
+
+  // The only kinds of links we can follow are 'reference search parameters'. This includes elements of type
+  // Reference and type canonical
+  if (!elements.every((elem) => elem.type === PropertyType.Reference || elem.type === PropertyType.canonical)) {
+    throw new OperationOutcomeError(badRequest('Invalid link path. Must return a path to a Reference type'));
+  }
+
+  // For each of the elements we found on the current resource, follow their various targets
+
+  const referenceElements = elements.filter((elem) => elem.type === PropertyType.Reference);
+  if (referenceElements.length > 0) {
+    results.push(...(await followReferenceElements(referenceElements, target, resourceCache, repo)));
+  }
+
+  const canonicalElements = elements.filter((elem) => elem.type === PropertyType.canonical);
+  if (canonicalElements.length > 0) {
+    results.push(...(await followCanonicalElements(canonicalElements, target, resourceCache, repo)));
+  }
+
+  return results;
+}
+
+async function followReferenceElements(
+  elements: TypedValue[],
+  target: GraphDefinitionLinkTarget,
+  resourceCache: Record<string, Resource>,
+  repo: Repository
```

Comment:
We try to use the convention that `repo` should always be the first argument to any server-side function requiring a `repo`.